Make tcx optional from StableMIR run macro and extend it to accept closures

celinval opened the pull request:
Change `run` macro to avoid sometimes unnecessary dependency on `TyCtxt`, and introduce `run_with_tcx` to capture use cases where `tcx` is required. Additionally, extend both macros to accept closures that may capture variables.

I've also modified the `internal()` method to make it safer, by accepting the type context to force the `'tcx` lifetime to match the context lifetime.

These are non-backward compatible changes, but they only affect internal APIs which are provided today as helper functions until we have a stable API to start the compiler.

celinval committed:
Remove tcx from SMIR run macro and accept closures

Simplify the `run` macro to avoid sometimes unnecessary dependency
on `TyCtxt`. Instead, users can use the new internal method `tcx()`.
Additionally, extend the macro to accept closures that may capture
variables.

These are non-backward compatible changes, but they only affect
internal APIs which are provided today as helper functions until we
have a stable API to start the compiler.

celinval commented:
r? @oli-obk

@oli-obk do you have time to review this one? Thanks

oli-obk commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -24,14 +24,49 @@ use std::ops::Index;
+/// # Panics
+///
+/// This function will panic if StableMIR has not been properly initialized.
 pub fn internal<'tcx, S: RustcInternal<'tcx>>(item: S) -> S::T {
     with_tables(|tables| item.internal(tables))
 }
 
+/// Retrieve the internal Rust compiler type context.
+///
+/// # Warning
+///
+/// This function is unstable, and it's behavior may change at any point.
+///
+/// # Panics
+///
+/// This function will panic if StableMIR has not been properly initialized.
+pub fn tcx<'tcx>() -> TyCtxt<'tcx> {
+    with_tables(|tables| tables.tcx)
+}
```
These two functions are unsound. They trivially allow you to create a `TyCtxt<'static>` or a `Ty<'static>` and move it outside the `run!` invocation in safe code.

The only way I see them becoming safe is to use the `with_tables` scheme of passing a closure that then gets the item as an argument. This way we can have control over how long the item lives

celinval replied:
Indeed. I think I prefer closing this PR and changing the `internal()` interface to take tcx as an argument. What do you think?

celinval replied:
I removed `tcx()` and added the type context as an argument to `internal()` to avoid the issue you just mentioned.

Even though the new solution is a bit hacky, I think its safer and more usable than the `with_tables` solution. Let me know what you think.

oli-obk replied:
Oh that's even better!

oli-obk reviewed (commented):
just some typos

celinval committed:
Remove tcx function and make `internal` fn safer

I added `tcx` argument to `internal` to force 'tcx to be the same
lifetime as TyCtxt. The only other solution I could think is to change
this function to be `unsafe`.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7548381047/job/20550307918) [(plain)](https://github.com/rust-lang/rust/commit/2564811e7b7464d4e2ab8cd5a826bf7c2f9f7d8b/checks/20550307918/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=celinval
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_8b86bf09-3f70-4b62-b490-1f3a70566bf3
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=cdccd60f270b3807aed036de7d84a2be9c78448d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_8b86bf09-3f70-4b62-b490-1f3a70566bf3
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_8b86bf09-3f70-4b62-b490-1f3a70566bf3
GITHUB_TRIGGERING_ACTOR=celinval
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/119833/merge
GITHUB_WORKFLOW_SHA=cdccd60f270b3807aed036de7d84a2be9c78448d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
 Documenting rustc_query_impl v0.0.0 (/checkout/compiler/rustc_query_impl)
 Documenting rustc_metadata v0.0.0 (/checkout/compiler/rustc_metadata)
 Documenting rustc_monomorphize v0.0.0 (/checkout/compiler/rustc_monomorphize)
 Documenting rustc_smir v0.0.0 (/checkout/compiler/rustc_smir)
error: `run` is both a function and a macro
    |
    |
296 | /// Prefer using [run] and [run_with_tcx] instead.
    |                   ^^^ ambiguous link
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
help: to link to the function, add parentheses
    |
    |
296 | /// Prefer using [run()] and [run_with_tcx] instead.
help: to link to the macro, add an exclamation mark
    |
    |
296 | /// Prefer using [run!] and [run_with_tcx] instead.

error: could not document `rustc_smir`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name rustc_smir compiler/rustc_smir/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'cfg()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=224cd140f870ed3a -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern rustc_abi=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_abi-ed39bb934b89e5bf.rmeta --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-00b6336307b8caec.rmeta --extern rustc_hir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_hir-de0ff62e76c3dbb6.rmeta --extern rustc_middle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_middle-87f10ef64bc278a1.rmeta --extern rustc_span=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_span-270d28216c71b709.rmeta --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-f5ef06ad65b368dd.rmeta --extern scoped_tls=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libscoped_tls-ed2db062fd97003f.rmeta --extern stable_mir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libstable_mir-d9a3640ddc85a8a0.rmeta --extern tracing=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libtracing-5f701b99da500b8b.rmeta --extern-html-root-url 'scoped_tls=https://docs.rs/scoped-tls/1.0.1/' --extern-html-root-url 'tracing=https://docs.rs/tracing/0.1.37/' -Zunstable-options --cfg=bootstrap --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=cfg(bootstrap)' '--check-cfg=cfg(parallel_compiler)' '--check-cfg=cfg(no_btreemap_remove_entry)' '--check-cfg=cfg(crossbeam_loom)' '--check-cfg=cfg(span_locations)' '--check-cfg=cfg(rustix_use_libc)' '--check-cfg=cfg(emulate_second_only_system)' '--check-cfg=cfg(windows_raw_dylib)' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.77.0-nightly (cdccd60f2 2024-01-16)' --document-private-items '-Arustdoc::private-intra-doc-links' --enable-index-page -Zunstable-options -Znormalize-docs --show-type-layout --generate-link-to-definition '-Zcrate-attr=warn(rust_2018_idioms)' --cfg=parallel_compiler --extern-html-root-url 'ena=https://docs.rs/ena/latest/'` (exit status: 1)
Build completed unsuccessfully in 0:01:12
  local time: Tue Jan 16 22:56:12 UTC 2024
  network time: Tue, 16 Jan 2024 22:56:12 GMT
##[error]Process completed with exit code 1.
```

</details>

celinval commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -190,35 +214,120 @@ where
-    ($args:expr, $tcx:ident, $callback:expr) => {{
+}
+
+/// Prefer using [run] and [run_with_tcx] instead.
```
```suggestion
/// Prefer using [run!] and [run_with_tcx] instead.
```

oli-obk commented:
@bors r+ rollup

bors commented:
:pushpin: Commit ac66570a6b93f6c57bf63865bf2a4da30eca8049 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk ac66570a6b93f6c57bf63865bf2a4da30eca8049 -->
<!-- homu: {"type":"Approved","sha":"ac66570a6b93f6c57bf63865bf2a4da30eca8049","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -24,11 +24,35 @@ use std::ops::Index;
+/// This function will panic if StableMIR has not been properly initialized.
+pub fn internal<'tcx, S: RustcInternal<'tcx>>(tcx: TyCtxt<'tcx>, item: S) -> S::T {
+    // The tcx argument ensures that the item won't outlive the type context.
+    let _ = tcx;
```
There can be multiple TyCtxt at the same time. You did have to use Lift to ensure that the item comes from the right TyCtxt.

oli-obk replied:
can we actually have them at the same time? Or just after each other? (though I guess the same issue applies there)

oli-obk replied:
But, that is not an issue here, because once the stable mir context shuts down, there is no more mapping of the stable mir type. You may get the wrong rustc type, but it will still be within the TyCtxt.

bjorn3 replied:
> can we actually have them at the same time?

Yes, you can a new TyCtxt and enter it and then with access to that TyCtxt it should be possible to create a new stable mir context and pass the TyCtxt you entered yourself as argument to internal(). This would allow the returned value outlive the TyCtxt of the stable mir context despite referencing values that are being freed with the rest of the stable mir context.

oli-obk replied:
Ok... I'm amazed our TLS doesn't panic, but just replace the old tls tcx with the new one and switch it back afterwards. I did not expect that.

celinval replied:
Just to make sure I understand, there could exist 2 or more `TyCtxt` in the same thread, and an item that was interned in one `TyCtxt` might outlive the type in a different one.

In that case, we should invoke `lift` using the `TyCtxt` that was passed to the `intern` function. Is that correct?

It also looks like not everything that implements `RustcInternal`, implements `Lift`, so the intern call will have to be done in the `RustcInternal` implementation for types that require so. 

If both facts are true, my plan is to mark the `RustcInternal` trait as `unsafe` and add `TyCtxt` as an argument to `RustcInternal::internal` as well. It will be up to the `RustcInternal` implementation to invoke `lift`.

celinval replied:
BTW, this change won't be small. Since the changes in this PR are now unrelated to this issue, can we go ahead and merge the current PR and I can create a follow up PR with the solution I described above. I can even revert adding `tcx` argument to the internal function if you prefer and add it in the next one.

celinval committed:
Revert changes to internal method for now

- Move fix to a separate PR

celinval commented:
@oli-obk I reverted all changes related to `internal()` from this PR as I'm going to create a separate PR for it. Please let me know if there is any other change you think should be done regarding this PR. Thanks!

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 6a573cbc60af8319e11cf0f0a578f8a12caa932d has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 6a573cbc60af8319e11cf0f0a578f8a12caa932d -->
<!-- homu: {"type":"Approved","sha":"6a573cbc60af8319e11cf0f0a578f8a12caa932d","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
